### PR TITLE
Changes for PR3 of BlameModeForDump

### PR DESCRIPTION
**Pull request by karanjitsingh**
This is the third and final PR of BlameModeForDump 
* TestHostLaunched payload is now a separate class (TestHostLaunchedPayload) instead of sending TestHostLaunchedEventArgs as payload
* Added CollectDump option check to EnableBlameDumpArgumentProcessor
* AcceptanceTest for dump file

### CODE REVIEW

File: src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/TestHostLaunchedEvent.cs [hunk 1]
```diff
@@ -43,7 +43,8 @@ public int TestHostProcessId
         /// Default constructor with default DataCollectionContext.
         /// DataCollectionContext with empty session signifies that is it irrelevent in the current context.
         /// </remarks>
-        public TestHostLaunchedEventArgs(int processId)
+        public TestHostLaunchedEventArgs(DataCollectionContext context, int processId)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.ObjectModel/DataCollector/Events/TestHostLaunchedEvent.cs]: Update summary.

File: test/vstest.console.UnitTests/Processors/EnableBlameArgumentProcessorTests.cs [hunk 2]
```diff
@@ -67,5 +76,57 @@ public void InitializeShouldCreateEntryForBlameInRunSettingsIfNotAlreadyPresent(
...
+                               .Returns(PlatformOperatingSystem.Windows);
+
+            this.mockEnvronment.SetupSequence(x => x.Architecture)
+                               .Returns(PlatformArchitecture.ARM)
+                               .Returns(PlatformArchitecture.ARM64);
+
+            for (var i = 0; i < 3; i++)
+            {
+                this.executor.Initialize("collectdump");
+                this.mockOutput.Verify(x => x.WriteLine(CommandLineResources.BlameCollectDumpNotSupportedForPlatform, OutputLevel.Warning));
```
Comment (smadala) [inline test/vstest.console.UnitTests/Processors/EnableBlameArgumentProcessorTests.cs]: Check runsettings.

**Review by smadala — APPROVED** [review]

**Comment by karanjitsingh** [comment]
@dotnet-bot test Windows_NT / Debug Build

**Comment by karanjitsingh** [comment]
@dotnet-bot test Windows_NT / Release Build

**Comment by karanjitsingh** [comment]
@dotnet-bot test this

**Comment by karanjitsingh** [comment]
@dotnet-bot test Windows_NT / Debug Build

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 3]
```diff
@@ -90,10 +90,10 @@ public void StartProcessDump(int processId, string dumpFileGuid, string testResu
         /// <returns>Arguments</returns>
         private static string BuildProcDumpArgs(int processId, string filename)
         {
+            // -accepteula: Accept EULA
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: elaborate

File: src/vstest.console/Processors/EnableBlameArgumentProcessor.cs [hunk 4]
```diff
@@ -15,6 +15,10 @@ namespace Microsoft.VisualStudio.TestPlatform.CommandLine.Processors
     using Microsoft.VisualStudio.TestPlatform.Common.Utilities;
     using Microsoft.VisualStudio.TestPlatform.CommandLine.Processors.Utilities;
     using System.Xml;
+    using Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces;
```
Comment (singhsarab) [inline src/vstest.console/Processors/EnableBlameArgumentProcessor.cs]: sort

File: src/vstest.console/Processors/EnableBlameArgumentProcessor.cs [hunk 5]
```diff
@@ -117,6 +135,23 @@ internal EnableBlameArgumentExecutor(IRunSettingsProvider runSettingsManager)
...
+                    this.environment.Architecture != PlatformArchitecture.ARM)
+                {
+                    isDumpEnabled = true;
+                }
+                else
+                {
+                    Output.Warning(false, CommandLineResources.BlameCollectDumpNotSupportedForPlatform);
+                }
+
+            }
```
Comment (singhsarab) [inline src/vstest.console/Processors/EnableBlameArgumentProcessor.cs]: nit: extra line

File: src/vstest.console/Processors/EnableBlameArgumentProcessor.cs [hunk 6]
```diff
@@ -162,7 +197,14 @@ public void Initialize(string argument)
             outernode.AppendChild(node);
             node.InnerText = resultsDirectory;
 
-            foreach(var item in dataCollectionRunSettings.DataCollectorSettingsList)
+            if (isDumpEnabled)
+            {
+                var dumpNode = XmlDocument.CreateElement("CollectDump");
```
Comment (singhsarab) [inline src/vstest.console/Processors/EnableBlameArgumentProcessor.cs]: >CollectDump [](start = 58, length = 11)

Make a const, multiple places

File: src/vstest.console/Processors/EnableBlameArgumentProcessor.cs [hunk 7]
```diff
@@ -162,7 +197,14 @@ public void Initialize(string argument)
...
+                outernode.AppendChild(dumpNode);
+                dumpNode.InnerText = "true";
```
Comment (singhsarab) [inline src/vstest.console/Processors/EnableBlameArgumentProcessor.cs]: >true [](start = 38, length = 4)

Why do we need inner text?

File: src/vstest.console/Resources/Resources.resx [hunk 8]
```diff
@@ -161,6 +161,9 @@
   <data name="BatchSizeRequired" xml:space="preserve">
     <value>The /BatchSize argument requires the size of the batch.  Example:  /BatchSize:10</value>
   </data>
+  <data name="BlameCollectDumpNotSupportedForPlatform" xml:space="preserve">
+    <value>CollectDump option for Blame is not supported for this platform.</value>
```
Comment (singhsarab) [inline src/vstest.console/Resources/Resources.resx]: Reviewed from Pratap ?

File: test/Microsoft.TestPlatform.AcceptanceTests/Microsoft.TestPlatform.AcceptanceTests.csproj [hunk 9]
```diff
@@ -21,6 +21,7 @@
     <PackageReference Include="Microsoft.TestPlatform.QTools.Assets">
       <Version>2.0.0</Version>
     </PackageReference>
+    <PackageReference Include="Procdump" Version="0.0.1" />
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.AcceptanceTests/Microsoft.TestPlatform.AcceptanceTests.csproj]: What package is this ?

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 10]
```diff
@@ -44,10 +44,11 @@ public string GetDumpFile()
...
             this.processHelper.WaitForProcessExit(this.procDumpProcess);
 
-            var dumpFile = Path.Combine(this.testResultsDirectory, this.dumpFileName);
-            if (this.fileHelper.Exists(dumpFile))
+            var dumpFiles = this.fileHelper.GetFiles(this.testResultsDirectory, this.dumpFileName + "*", SearchOption.TopDirectoryOnly);
+            if (dumpFiles.Length > 0)
             {
-                return dumpFile;
+                // Dump files can never be more than 1 because procdump will generate single file, but GetFiles function returns an array
+                return dumpFiles[0];
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: probably log if there are multiple

File: test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/BlameCollectorTests.cs [hunk 11]
```diff
@@ -210,12 +210,20 @@ public void TriggerTestHostLaunchedHandlerShouldCatchAllExceptions()
                 this.context);
 
             // Make StartProcessDump throw exception
-            this.mockProcessDumpUtility.Setup(x => x.StartProcessDump(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<string>()))
-                                       .Throws(new Exception("start process failed"));
+            var ex = new Exception("start process failed");
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/BlameCollectorTests.cs]: 2 separate tests.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 12]
```diff
@@ -44,10 +44,17 @@ public string GetDumpFile()
...
+            if (dumpFiles.Length > 0)
             {
-                return dumpFile;
+                // Log to diagnostics if multiple files just in case
+                if (dumpFiles.Length != 1)
+                {
+                    EqtTrace.Warning("ProcessDumpUtility.GetDumpFile: Multiple dump files found.");
+                }
+
+                return dumpFiles[0];
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: if clean up of dumps did not happen, by returning dumpfiles[0], you could be returning stale dump?

**Pull request merged**